Flaky test: awslogs tests failing with ResourceAlreadyExistsException

Not sure if related to similar errors, but there were some fixes for race-conditions in the past (see https://github.com/moby/moby/issues/22963, and https://github.com/moby/moby/pull/22985)

Seen this failing a couple of times: 

Failure on RS1 (but I may have seen this on other platforms); https://ci.docker.com/public/blue/rest/organizations/jenkins/pipelines/moby/branches/PR-39855/runs/1/nodes/39/log/?start=0

```
time="2019-09-03T10:15:19Z" level=info msg="Trying to get region from EC2 Metadata"
time="2019-09-03T10:15:19Z" level=info msg="Log stream already exists" errorCode=ResourceAlreadyExistsException logGroupName= logStreamName= message= origError="<nil>"
--- FAIL: TestLogBlocking (0.02s)
    cloudwatchlogs_test.go:313: Expected to be able to read from stream.messages but was unable to
time="2019-09-03T10:15:19Z" level=error msg=Error
time="2019-09-03T10:15:19Z" level=error msg="Failed to put log events" errorCode=InvalidSequenceTokenException logGroupName=groupName logStreamName=streamName message="use token token" origError="<nil>"
time="2019-09-03T10:15:19Z" level=error msg="Failed to put log events" errorCode=DataAlreadyAcceptedException logGroupName=groupName logStreamName=streamName message="use token token" origError="<nil>"
time="2019-09-03T10:15:19Z" level=info msg="Data already accepted, ignoring error" errorCode=DataAlreadyAcceptedException logGroupName=groupName logStreamName=streamName message="use token token"
FAIL
coverage: 78.2% of statements
FAIL	github.com/docker/docker/daemon/logger/awslogs	0.631s
```


Interesting bit in that case is that the test that was marked failing looks to be using a `mockClient` (so should not even be connecting to AWS?) https://github.com/moby/moby/blob/e78a3dca21ea9ceb8d7a5e0252e94ee7ad05eadd/daemon/logger/awslogs/cloudwatchlogs_test.go#L288-L322

For other tests, we seem to be reusing the same `groupName` and `logStreamName`, so I suspect this is just a race condition (and AWS doesn't like attempting to create duplicates); https://github.com/moby/moby/blob/e78a3dca21ea9ceb8d7a5e0252e94ee7ad05eadd/daemon/logger/awslogs/cloudwatchlogs_test.go#L29-L36


Should we add a unique suffix to the `groupName` and `logStreamName` in the tests? (e.g., use the test name as suffix? `groupName+t.Name()`)

--- a/daemon/logger/awslogs/cloudwatchlogs_test.go
+++ b/daemon/logger/awslogs/cloudwatchlogs_test.go
@@ -24,7 +24,6 @@ import (
 	"github.com/docker/docker/dockerversion"
 	"gotest.tools/assert"
 	is "gotest.tools/assert/cmp"
-	"gotest.tools/skip"
 )
 
 const (
@@ -286,8 +285,10 @@ func TestLogClosed(t *testing.T) {
 	}
 }
 
+// TestLogBlocking tests that the Log method blocks appropriately when
+// non-blocking behavior is not enabled.  Blocking is achieved through an
+// internal channel that must be drained for Log to return.
 func TestLogBlocking(t *testing.T) {
-	skip.If(t, runtime.GOOS == "windows", "FIXME: test is flaky on Windows. See #39857")
 	mockClient := newMockClient()
 	stream := &logStream{
 		client:   mockClient,
@@ -301,18 +302,20 @@ func TestLogBlocking(t *testing.T) {
 		err := stream.Log(&logger.Message{})
 		errorCh <- err
 	}()
+	// block until the goroutine above has started
 	<-started
 	select {
 	case err := <-errorCh:
 		t.Fatal("Expected stream.Log to block: ", err)
 	default:
-		break
 	}
+	// assuming it is blocked, we can now try to drain the internal channel and
+	// unblock it
 	select {
-	case <-stream.messages:
-		break
-	default:
+	case <-time.After(10 * time.Millisecond):
+		// if we're unable to drain the channel within 10ms, something seems broken
 		t.Fatal("Expected to be able to read from stream.messages but was unable to")
+	case <-stream.messages:
 	}
 	select {
 	case err := <-errorCh:

